src/graph/graph.c

```diff
@@ -25,250 +27,82 @@ void _Graph_LeaveCriticalSection(Graph *g) {
-size_t _Graph_NodeCap(const Graph *g) {
-    return g->nodes->itemCap;
-}
-
-// Resize graph's node array to contain at least n nodes.
-void _Graph_AddNodes(Graph *g, size_t n, DataBlockIterator **it) {
-    DataBlock_AddItems(g->nodes, n, it);
-}
-
-// Resize graph's edge array to contain at least n edges.
-void _Graph_AddEdges(Graph *g, size_t n, DataBlockIterator **it) {
-    DataBlock_AddItems(g->edges, n, it);
+    GrB_Matrix m = g->_relations_map[relation_idx];
+    _Graph_ResizeMatrix(g, m);
+    return m; 
 }
 
-/* Removes a single entry from given matrix. */
-void _Graph_ClearMatrixEntry(Graph *g, GrB_Matrix M, GrB_Index src, GrB_Index dest) {
-    GrB_Vector mask;
-    GrB_Index nrows = Graph_NodeCount(g);
-    GrB_Vector_new(&mask, GrB_BOOL, nrows);
-    GrB_Vector_setElement_BOOL(mask, true, dest);
-
-    GrB_Vector col;
-    GrB_Vector_new(&col, GrB_BOOL, nrows);
-
-    GrB_Descriptor desc;
-    GrB_Descriptor_new(&desc);
-    GrB_Descriptor_set(desc, GrB_OUTP, GrB_REPLACE);
-    GrB_Descriptor_set(desc, GrB_MASK, GrB_SCMP);
-
-    // Extract column src_id.
-    GrB_Col_extract(col, mask, NULL, M, GrB_ALL, nrows, src, desc);
-    GrB_Col_assign(M, NULL, NULL, col, GrB_ALL, nrows, src, NULL);
-
-    GrB_Descriptor_free(&desc);
-    GrB_Vector_free(&col);
-    GrB_Vector_free(&mask);
+// Create a new mapping matrix M,
+// M[I,J] holds the edge ID connecting node J to I (CSC format)
+// assuming _relations_map[K] holds mapping for relation K.
+void _Graph_AddRelationMap(Graph *g) {
+    GrB_Matrix mapper;
+    GrB_Info res = GrB_Matrix_new(&mapper, GrB_UINT64, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));
+    assert(res == GrB_SUCCESS);
+    g->_relations_map = array_append(g->_relations_map, mapper);
 }
 
-/* Initialize edge and store it within graph's edge hashtable
- * edge key within the hashtable is composed of:
- * 1. Edge relation ID
- * 2. Edge source node ID
- * 3. Edge destination node ID */
-int _Graph_InitEdge(Graph *g, Edge *e, EdgeID id, Node *src, Node *dest, int r) {
-    // Insert only if edge not already in hashtable.
-    Edge *edge;
-    EdgeDesc lookupKey = {src->id, dest->id, r};
-    HASH_FIND(hh, g->_edgesHashTbl, &lookupKey, sizeof(EdgeDesc), edge);
+// Locates edge connecting src to destination.
+Entity* _Graph_GetEdgeConnectingNodes(const Graph *g, NodeID src, NodeID dest, int r) {
+    assert(g && src < Graph_RequiredMatrixDim(g) && dest < Graph_RequiredMatrixDim(g) && r < Graph_RelationTypeCount(g));
 
-    if(edge) {
-        /* TODO: Currently we only support a single edge of type T
-         * between tow nodes. */
-        return 0;
-    }
+    // relation map, maps (src, dest, r) to edge id.
+    GrB_Matrix relationMap = _Graph_GetRelationMap(g, r);
 
-    // Set edge composite ID.
-    e->id = id;
-    Edge_SetSrcNode(e, src);
-    Edge_SetDestNode(e, dest);
-    Edge_SetRelationID(e, r);
+    EdgeID edgeId = 0;
+    GrB_Matrix_extractElement_UINT64(&edgeId, relationMap, dest, src);
```

Comment:
We should be able to avoid off-by-one weirdness with something like:
```
GrB_Info res = GrB_Matrix_extractElement_UINT64...
if (res == GrB_NO_VALUE) return NULL;
```

Reply:
👍